api/src/main/java/org/openmrs/api/impl/TagServiceImpl.java

```diff
@@ -0,0 +1,201 @@
+import org.openmrs.Tag;
+import org.openmrs.api.TagService;
+import org.openmrs.api.context.Context;
+import org.openmrs.api.db.TagDAO;
+import org.openmrs.util.OpenmrsClassLoader;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.ArrayList;
+
+@Transactional(readOnly = true)
+public class TagServiceImpl extends BaseOpenmrsService implements TagService {
+	
+	protected final Log log = LogFactory.getLog(this.getClass());
+	
+	private TagDAO dao;
+	
+	/**
+	 * Injected in moduleApplicationContext.xml
+	 */
+	public void setDao(TagDAO dao) {
+		this.dao = dao;
+	}
+	
+	/**
+	 * @see TagService#getTagByUuid(String)
+	 */
+	@Override
+	public Tag getTagByUuid(String uuid) {
+		return dao.getTagByUuid(uuid);
+	}
+	
+	/**
+	 * @see TagService#saveTag(Tag)
+	 */
+	@Override
+	@Transactional(readOnly = false)
+	public Tag saveTag(Tag tag) {
+		if (getObject(toClass(tag.getObjectType()), tag.getObjectUuid()) == null) {
+			log.warn("Object does not exist in the the database");
+		}
+		return dao.saveTag(tag);
+	}
+	
+	/**
+	 * @see TagService#purgeTag(Tag)
+	 */
+	@Override
+	@Transactional(readOnly = false)
+	public void purgeTag(Tag tag) {
+		dao.deleteTag(tag);
+	}
+	
+	/**
+	 * @see TagService#removeTag(OpenmrsObject, String)
+	 */
+	@Override
+	@Transactional(readOnly = false)
+	public void removeTag(OpenmrsObject openmrsObject, String tag) {
+		Tag tag1 = dao.getTag(openmrsObject, tag);
+		if (tag1 == null) {
+			log.warn("Tag does not exist");
+		} else {
+			purgeTag(tag1);
+		}
+	}
+	
+	/**
+	 * @see TagService#getAllTags()
+	 */
+	@Override
+	public List<String> getAllTags() {
+		return dao.getAllTags();
+	}
+	
+	/**
+	 * @see TagService#getTag(Integer)
+	 */
+	@Override
+	public Tag getTag(Integer tagId) {
+		return dao.getTag(tagId);
+	}
+	
+	/**
+	 * @see TagService#getTags(String)
+	 */
+	@Override
+	public List<Tag> getTags(String searchPhrase) {
+		return dao.getTags(searchPhrase);
+	}
+	
+	/**
+	 * Retrieves an Object with given object uuid and object type from the Openmrs database.
+	 */
+	private <T extends OpenmrsObject> T getObject(Class<T> objectType, String objectUuid) {
+		Object object;
+		object = dao.getObject(objectType, objectUuid);
+		return (T) object;
+	}
+	
+	/**
+	 * @see TagService#addTag(OpenmrsObject, String)
+	 */
+	@Override
+	@Transactional(readOnly = false)
+	public Tag addTag(OpenmrsObject openmrsObject, String tag) {
+		if (hasTag(openmrsObject, tag)) {
+			log.warn("duplicate Tag for " + openmrsObject);
+		} else {
+			Tag tag1 = new Tag(tag, openmrsObject.getUuid(), openmrsObject.getClass().getName());
+			return saveTag(tag1);
+		}
+		return null;
+	}
+	
+	/**
+	 * Validation Method which returns false if u try to add a duplicate tag to an OpenmrsObject
+	 */
+	public boolean hasTag(OpenmrsObject openmrsObject, String tag) {
+		Tag tag1 = dao.getTag(openmrsObject, tag);
```

Comment:
You can combine these into a single line as below:

`return dao.getTag(openmrsObject, tag) != null;`